blog/_posts/2023-01-16-wsl2spackapptainer.md

```diff
@@ -0,0 +1,284 @@
+
+### Install waLBerla
+
+There isn't a package for waLBerla, but it is a standard CMake package, so should be easy for Spack to handle.
+
+```bash
+spack create https://i10git.cs.fau.de/walberla/walberla/-/archive/v6.1/walberla-v6.1.tar.bz2
+```
+
+![Basic recipe](/images/blog/wsl2spackapptainer/initialrecipe.png)
+Now this isn't a perfect package file, as it doesn't have any dependencies defined, but it's a very nice starting point, given how little effort we've had to put in to make it.  Let's give that a try and see what fails, if anything:
+
+```bash
+spack install walberla
+```
+
+This goes through and installs all the things it thinks it needs to build a generic CMake package, then fails:
+
+```
+Wrapper.h:68:10: fatal error: mpi.h: No such file or directory
+```
+
+Ah okay, it needs MPI to be installed. Edit the recipe (`spack edit walberla`) so we can add a dependency on mpi where it says in the file:
+
+```python
+depends_on("mpi")
+```
+
+How did I know it was a package called "mpi" that I needed to add here?  I actually just did `spack info hdf5` to take a look at another package I knew probably had a similar requirement, and learnt from that.
+
+Retry the install (`spack install walberla`) and it fails again, this time with a new error:
+
+```
+987     g++: fatal error: Killed signal terminated program cc1plus
+```
+
+What's happened here is that we've run out of memory, since by default, WSL2 gets no more than 50%/8Gbytes of RAM, and 2Gbytes of swap.  To increase this, create a file in your Windows home directory (e.g `C:\Users\exuser`) called .wslconfig containing:
+
+```bash
+[wsl2]
+# Limits VM memory to use no more than 12 GB, defaults to 50% of ram
+memory=12GB
+# Sets the amount of swap storage space to 16GB, default is 25% of available RAM
+swap=16GB
+```
+
+In my case, this is just enough to let it build successfully, but you may want different figures for your system.  It appears to need a total of ~15Gbytes to run all in memory, but that wasn't an option on the 16Gbyte machine I tested on.  Pick figures that work for you.
+
+You need to restart wsl at this point, so in a Windows command prompt:
+
+```bash
+wsl --shutdown
+```
+
+Start your Ubuntu session back up (I just clicked the Ubuntu entry in the Start Menu), reload spack into your environment, and build waLBerla again:
+
+```bash
+. spack/share/spack/setup-env.sh
+spack install walberla
+```
+
+Spack will only rebuild what it needs to, so it should skip straight to building waLBerla, and this time it will succeed.
+
+### Why I'm off piste...
+
+This is the point I realise that what I thought I understood about how espressomd integrated with waLBerla was total nonsense, and I'd been digging an interesting shaped hole all this time.  The integration isn't present in the main fork, and when it is included, it pulls in its own copy of waLBerla rather than uses an existing build version!  Ah well, at least we know we've got all the pieces in places for this to work...
```

Comment:
```suggestion
This is the point I realise that what I thought I understood about how espressomd integrated with waLBerla was total nonsense, and I'd been digging an interesting shaped hole all this time.  The integration isn't present in the main fork, and when it is included, it pulls in its own copy of waLBerla rather than using an existing build version!  Ah well, at least we know we've got all the pieces in places for this to work...
```